File: 数据结构/编程/排序/hell_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>


using namespace std;

//选择排序，选择最小的放在前面
//插入排序，就是整体把整个队列向后移动，然后插入到合适的位置

void hell_sort(vector<int>& vec, int first, int last)
{
    for(int i = first; i<last; i++)
    {
        for(int N = (last-first)+1; N>=1; N/=2)
        for(int j = first+1; j<=last; j+=N)
        {
            if(vec[i]>vec[j])
                swap(vec[i], vec[j]);
        }
    }
}

int main()
{
    vector<int> in = {4,5,6,7,8,9,10};//{10, 9, 8,7,6,5,4,3,2,1,0};
    vector<int> in1 = {10, 9, 8,7,6,5,4,3,2,1,0};
    //sort(in.begin(), in.end());
    hell_sort(in1, 0, in1.size()-1);
    for(auto item:in1)
        cout<<item<<endl;

    return 0;
}

File: alg/src/xsort/hell_sort.go
package xsort

/*
#include <iostream>
#include <vector>

using namespace std;

void hell_sort(vector<int>&vec, int first, int last)
{
    for(int i = first; i<last; ++i)
    {

        for(int N = (last-first+1)/2; N>=1; N/=2)
        {
            int tmp_min_index = i;
            for(int j = i+N; j<=last; j+=N)
            {
                if(vec[i] > vec[j])
                {
                    tmp_min_index = j;
                }
            }

            swap(vec[i], vec[tmp_min_index]);
        }

    }
}

int main()
{
    vector<int> vec = {9,8,7,6,5,4,3,2,1,0};
    hell_sort(vec, 0, vec.size()-1);
    for(auto item:vec)
    {
        cout<<item<<endl;
    }
    return 0;
}
*/
